test/src/test/java/jenkins/security/AgentSecretActionTest.java

```
@@ -0,0 +1,85 @@
+import org.kohsuke.stapler.StaplerRequest2;
+import org.kohsuke.stapler.StaplerResponse2;
+
+public class AgentSecretActionTest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    private DumbSlave agent;
+
+    private StaplerRequest2 req;
+    private StaplerResponse2 rsp;
+    private StringWriter stringWriter;
+    private PrintWriter writer;
+
+    private String agentUrl;
+
+    @Before
+    public void setUp() throws Exception {
+        agent = new DumbSlave("test-agent", "/home/jenkins", new hudson.slaves.JNLPLauncher());
+        req = mock(StaplerRequest2.class);
+        rsp = mock(StaplerResponse2.class);
+        stringWriter = new StringWriter();
+        writer = new PrintWriter(stringWriter);
+        when(rsp.getWriter()).thenReturn(writer);
+        agentUrl = "computer/test-agent/agent-secret/";
+    }
+
+    @Test
+    public void testGetSecretWithValidPermissions() throws Exception {
+        String expectedSecret = ((SlaveComputer) agent.getComputer()).getJnlpMac();
+
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        MockAuthorizationStrategy authStrategy = new MockAuthorizationStrategy()
+                .grant(Computer.CONNECT).everywhere().to("test-user");
+        j.jenkins.setAuthorizationStrategy(authStrategy);
+        JenkinsRule.WebClient webClient = j.createWebClient();
+        webClient.login("test-user");
+        String response = webClient.goTo(agentUrl, "text/plain")
+                .getWebResponse().getContentAsString();
+
+        assertEquals(expectedSecret, response);
+
+    }
+
+    @Test
+    public void testGetSecretWithoutPermissions() throws Exception {
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(new MockAuthorizationStrategy());
+
+        JenkinsRule.WebClient webClient = j.createWebClient();
+        webClient.login("test-user");
```

You will need to grant the user `test-user` at least the permission `Jenkins.READ` . The test fails early because of this missing permission.